src/ansi-c/c_typecheck_code.cpp

```diff
@@ -842,35 +843,158 @@ void c_typecheck_baset::typecheck_dowhile(code_dowhilet &code)
   }
 }
 
+void c_typecheck_baset::typecheck_spec_assigns_condition(exprt &condition)
+{
+  // compute type
+  typecheck_expr(condition);
+
+  // non-fatal warnings
+  if(has_subexpr(condition, [&](const exprt &subexpr) {
+       return can_cast_expr<side_effect_expr_function_callt>(subexpr);
+     }))
+  {
+    // Remark: we allow function calls without further checks for now because
+    // they are mandatory for some applications.
+    // The next step must be to check that the called functions have a contract
+    // with an empty assigns clause and that they indeed satisfy their contract
+    // using a CI check.
+    warning().source_location = condition.source_location();
+    warning()
+      << "function with possible side-effect called in target's condition"
+      << eom;
+  }
+
+  // fatal errors
+  bool must_throw = false;
+
+  if(condition.type().id() == ID_empty)
+  {
+    must_throw = true;
+    error().source_location = condition.source_location();
+    error() << "void-typed expressions not allowed as assigns clause conditions"
+            << eom;
+  }
+
+  if(has_subexpr(condition, [&](const exprt &subexpr) {
+       return can_cast_expr<side_effect_exprt>(subexpr) &&
+              !can_cast_expr<side_effect_expr_function_callt>(subexpr);
+     }))
+  {
+    must_throw = true;
+    error().source_location = condition.source_location();
+    error() << "side-effects not allowed in assigns clause conditions" << eom;
+  }
```

Comment:
How about making this an `else if` of the earlier `side_effect_expr_function_callt` check?

Reply:
Since these cases are not exclusive (as you can  have both a `side_effect_expr_function_callt` and `side_effect_exprt` that is not a `side_effect_expr_function_callt` in the same expression) we need to keep these checks independent.

Reply:
Related question: is there already a function such as 

```
iterator<exprt &> iter_subexpr(
  const exprt &expr,
  const std::function<bool(const exprt &)> &pred),
  traversal_ordert order
)
```

That would let me iterate (in some chosen or default order) on subexpressions that match a given predicate ?

Reply:
I don't think we have that in full, but at least the iteration can be done using one of the iterators defined in `util/expr_iterator.h`. This is used in several places, you might take a look at `util/rename_symbol.cpp` for an example.